parcels/interaction/base_hash.py

```diff
@@ -0,0 +1,167 @@
+from abc import ABC, abstractmethod
+
+import numpy as np
+
+
+class BaseHashNeighborSearch(ABC):
+    def find_neighbors_by_coor(self, coor):
+        '''Get the neighbors around a certain location.
+
+        :param coor: Numpy array with [lat, long, depth].
+        :returns List of particle indices.
+        '''
+        coor = coor.reshape(3, 1)
+        hash_id = self._values_to_hashes(coor)[0]
+        return self._find_neighbors(hash_id, coor)
+
+    def find_neighbors_by_idx(self, particle_idx):
+        '''Get the neighbors around a certain particle.
+
+        Mainly useful for Structure of Array (SoA) datastructure
+
+        :param particle_idx: index of the particle (SoA).
+        :returns List of particle indices
+        '''
+        hash_id = self._particle_hashes[particle_idx]
+        coor = self._values[:, particle_idx].reshape(3, 1)
+        return self._find_neighbors(hash_id, coor)
+
+    @abstractmethod
+    def _find_neighbors(self, hash_id, coor):
+        raise NotImplementedError
+
+    def consistency_check(self):
+        '''See if all values are in their proper place.
+
+        Only used for debugging purposes.
+        '''
+        active_idx = self.active_idx
+        if active_idx is None:
+            active_idx = np.arange(self._values.shape[1])
+
+        for idx in active_idx:
+            cur_hash = self._particle_hashes[idx]
+            hash_idx = self._hash_idx[idx]
+            assert self._hashtable[cur_hash][hash_idx] == idx
+
+        n_idx = 0
+        for idx_array in self._hashtable.values():
+            for idx in idx_array:
+                assert idx in active_idx
+            n_idx += len(idx_array)
+        assert n_idx == len(active_idx)
+        cur_hashes = self._values_to_hashes(self._values[:, active_idx])
+        assert np.all(cur_hashes == self._particle_hashes[active_idx])
+
+    def update_values(self, new_values, new_active_mask=None):
```

Comment:
very nice function, also the use of the comments is really useful here. I like it.